### PR TITLE
Streamline features with proto

### PR DESCRIPTION
**Pull request by ad-47**
Highlites:
- Pagination & total_pages support on Get/Find experimentRun
- Code Versioning, CodeVersion for each service
- ExperimentRun parent mapping
- Improved test cases
- Log4j logger implementation etc.

### CODE REVIEW

File: src/test/java/com/mitdbg/modeldb/mongoTest1Config.yaml [hunk 1]
```diff
@@ -4,18 +4,117 @@ grpcServer:
...
+        background: true
+        drop_if_exists: false
+        columns:
+         - name: comment_id
+           ordering: asc
+
+cloudconfig:
+  cloudCredentials: access-secret #for s3 its access-secret, maybe set these when building the image?, for gcp?, for nfs?
+  cloudType: S3 #S3, GCP, NFS
+  cloudBucketName: verta-artifacts # if nfs then root dir. Note: bucket needs to exist alrady
```
Comment (ravishetye) [inline src/test/java/com/mitdbg/modeldb/mongoTest1Config.yaml]: let un not have any verta specific mention in here. You can keep this empty (with just the comment)

File: src/test/java/com/mitdbg/modeldb/mongoTest2Config.yaml [hunk 2]
```diff
@@ -4,18 +4,117 @@ grpcServer:
...
+        background: true
+        drop_if_exists: false
+        columns:
+         - name: comment_id
+           ordering: asc
+
+cloudconfig:
+  cloudCredentials: access-secret #for s3 its access-secret, maybe set these when building the image?, for gcp?, for nfs?
+  cloudType: S3 #S3, GCP, NFS
+  cloudBucketName: verta-artifacts # if nfs then root dir. Note: bucket needs to exist alrady
```
Comment (ravishetye) [inline src/test/java/com/mitdbg/modeldb/mongoTest2Config.yaml]: no verta specific values

File: src/test/java/com/mitdbg/modeldb/setProjectReadMe.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
+# Project Title
```
Comment (ravishetye) [inline src/test/java/com/mitdbg/modeldb/setProjectReadMe.md]: add another header in the readme , that this is just the sample for testing project readme's
e.g.
# This is a test-case

File: src/main/java/com/mitdbg/modeldb/project/ProjectServiceImpl.java [hunk 4]
```diff
@@ -686,33 +740,65 @@ public void getProjects(
...
+   * enabled, it returns all the projects from the database.
+   *
+   * @param GetProjects request, GetProjects.Response response
+   * @return void
+   */
+  @Override
+  public void getPublicProjects(
+      GetPublicProjects request, StreamObserver<GetPublicProjects.Response> responseObserver) {
+    try {
+      List<Project> projectList = projectDAO.getPublicProjects(request.getUserId());
```
Comment (ravishetye) [inline src/main/java/com/mitdbg/modeldb/project/ProjectServiceImpl.java]: this should call getProjects

File: README.md [hunk 5]
```diff
@@ -31,6 +31,11 @@ Run gRPC server use following command
 ```
 3) mvn exec:java -Dexec.mainClass=com.mitdbg.modeldb.App
 ```
+### Running Tests
+
+The tests require mongo to be running. Set the host and port in config.yaml.
+
+Also set the VERTA_MODELDB_CONFIG environment variable to point to config.yaml
```
[the remark this answers is not in the document]
Comment (ad-47) [reply README.md]: Fixed

File: src/main/java/com/mitdbg/modeldb/App.java [hunk 6]
```diff
@@ -232,32 +259,33 @@ private static void createNewMongoUser(MongoDatabase database, String username,
...
+        for (Document roleDoc : rolesList) {
+          if (roleDoc.getString("role").equalsIgnoreCase("dbOwner")) {
+            doesAdminUserExist = true;
+            break;
           }
-          break;
         }
+        if (!doesAdminUserExist) {
+          throw new ModelDBException(
+              "Given user don't have a 'dbOwner' access control, system aspect the config user has 'dbOwner' access control");
```
[the remark this answers is not in the document]
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/App.java]: Done

File: src/main/java/com/mitdbg/modeldb/ModelDBConstants.java [hunk 7]
```diff
@@ -4,7 +4,7 @@
 
   // Default Credential
   String DEFAULT_USERNAME = "vertaMDBAdmin";
-  String DEFAULT_PASSWORD = "vertaMDBAdmin@123";
+  String DEFAULT_PASSWRD = "vertaMDBAdmin@123";
```
[the remark this answers is not in the document]
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/ModelDBConstants.java]: Fixed

File: src/main/java/com/mitdbg/modeldb/ModelDBUtils.java [hunk 8]
```diff
@@ -83,13 +87,26 @@ private static String getURLEncoderString(String inputString)
     return URLEncoder.encode(inputString, "UTF-8");
   }
 
+  public static String readFile(String filePath) throws IOException {
```
[the remark this answers is not in the document]
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/ModelDBUtils.java]: Done

File: src/main/java/com/mitdbg/modeldb/ModelDBUtils.java [hunk 9]
```diff
@@ -98,12 +115,53 @@ public static String convertToProjectShortName(String projectName) {
               .build();
       throw StatusProto.toStatusRuntimeException(status);
     }
-    if (projectShortName.length() > ModelDBConstants.NAME_MAX_LENGTH) {
+    if (projectShortName.length() > 42) {
```
[the remark this answers is not in the document]
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/ModelDBUtils.java]: Fixed

File: src/main/java/com/mitdbg/modeldb/artifactStore/ArtifactStoreDAORdbImpl.java [hunk 10]
```diff
@@ -74,4 +80,14 @@ public Boolean deleteStoreKeyMapping(String cloudStorageKey) {
...
     }
   }
+
+  @Override
+  public GetUrlForArtifact.Response getUrlForArtifact(String path, String method)
+      throws ModelDBException {
+    String errorMessage = "Method not implemented";
+    LOGGER.log(Level.WARNING, errorMessage);
+    Status status =
+        Status.newBuilder().setCode(Code.UNIMPLEMENTED_VALUE).setMessage(errorMessage).build();
```
[the remark this answers is not in the document]
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/artifactStore/ArtifactStoreDAORdbImpl.java]: Done.
Reference URL: https://github.com/VertaAI/modeldb-backend/issues/10

File: src/main/java/com/mitdbg/modeldb/databaseServices/DocumentService.java [hunk 11]
```diff
@@ -1,14 +1,15 @@
...
 import com.google.protobuf.MessageOrBuilder;
 
+import java.util.List;
+import java.util.Map;
+
 public interface DocumentService {
 
   /*
-   * Check availability of given collection name in database, if is not exist then create collection
+   * Check availability of given collection name in database, if does not exist then create collection
```
[the remark this answers is not in the document]
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/databaseServices/DocumentService.java]: Done

File: src/main/java/com/mitdbg/modeldb/experiment/ExperimentDAOMongoImpl.java [hunk 12]
```diff
@@ -15,33 +10,50 @@
...
+  private static final Logger LOGGER = LogManager.getLogger(ExperimentDAOMongoImpl.class);
+  private String projectEntity;
+  private String experimentEntity;
+  private String experimentRunEntity;
+  private DocumentService documentService;
+  private ExperimentRunDAO experimentRunDAO;
 
-  public ExperimentDAOMongoImpl(DocumentService documentService) {
+  public ExperimentDAOMongoImpl(
+      DocumentService documentService, ExperimentRunDAO experimentRunDAO) {
```
[the remark this answers is not in the document]
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/experiment/ExperimentDAOMongoImpl.java]: Yes, it is not used. Removed.

File: src/main/java/com/mitdbg/modeldb/project/ProjectDAOMongoImpl.java [hunk 13]
```diff
@@ -1,46 +1,64 @@
...
 import com.google.rpc.Code;
 import com.google.rpc.Status;
 import com.mitdbg.modeldb.App;
+import com.mitdbg.modeldb.Experiment;
+import com.mitdbg.modeldb.ExperimentRun;
 import com.mitdbg.modeldb.KeyValue;
 import com.mitdbg.modeldb.ModelDBConstants;
 import com.mitdbg.modeldb.ModelDBUtils;
 import com.mitdbg.modeldb.Project;
+import com.mitdbg.modeldb.ProjectVisibility;
```
[the remark this answers is not in the document]
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/project/ProjectDAOMongoImpl.java]: Removed

File: src/test/java/com/mitdbg/modeldb/mongoTest1Config.yaml (hunk 1 above)
Comment (ad-47) [reply src/test/java/com/mitdbg/modeldb/mongoTest1Config.yaml]: Fixed

File: src/test/java/com/mitdbg/modeldb/mongoTest2Config.yaml (hunk 2 above)
Comment (ad-47) [reply src/test/java/com/mitdbg/modeldb/mongoTest2Config.yaml]: Fixed

File: src/main/java/com/mitdbg/modeldb/project/ProjectServiceImpl.java (hunk 4 above)
Comment (ad-47) [reply src/main/java/com/mitdbg/modeldb/project/ProjectServiceImpl.java]: Done

File: src/test/java/com/mitdbg/modeldb/setProjectReadMe.md (hunk 3 above)
Comment (ad-47) [reply src/test/java/com/mitdbg/modeldb/setProjectReadMe.md]: Done

**Pull request merged**